nvtabular/ops.py

```diff
@@ -1242,30 +1239,32 @@ def op_logic(self, gdf: cudf.DataFrame, target_columns: list, stats_context={}):
                 new_gdf[new_col] = new_gdf[new_col].astype(self.dtype, copy=False)
         return new_gdf
 
-    def get_emb_sz(self, encoders, cat_names):
-        # sorted key required to ensure same sort occurs for all values
-        ret_list = [
-            (n, self.def_emb_sz(encoders, n))
-            for n in sorted(cat_names, key=lambda entry: entry.split("_")[0])
-        ]
-        return ret_list
 
-    def emb_sz_rule(self, n_cat: int) -> int:
-        return min(16, round(1.6 * n_cat ** 0.56))
+def get_embedding_order(cat_names):
+    """ Returns a consistent sorder order for categorical variables
 
-    def def_emb_sz(self, classes, n, sz_dict=None):
-        """Pick an embedding size for `n` depending on `classes` if not given in `sz_dict`.
-        """
-        sz_dict = sz_dict if sz_dict else {}
-        n_cat = classes[n]
-        sz = sz_dict.get(n, int(self.emb_sz_rule(n_cat)))  # rule of thumb
-        self.embed_sz[n] = sz
-        return n_cat, sz
-
-    def get_embeddings(self, encoders, cat_names):
-        embeddings = {}
-        for col in sorted(cat_names):
-            path = encoders[col]
-            num_rows, _, _ = cudf.io.read_parquet_metadata(path)
-            embeddings[col] = (num_rows, self.emb_sz_rule(num_rows))
-        return embeddings
+    Parameters
+    -----------
+    cat_names : list of str
+        names of the categorical columns
+    """
+    return sorted(cat_names, key=lambda entry: entry.split("_")[0])
+
+
+def get_embedding_size(encoders, cat_names):
```

Comment:
This comment might be out of the scope of this refactor, but we should try to make this function cleaner and clearer.

Right now, the function accepts a `dict` mapping names to cardinalities, along with a list of names. Then these are both used to simply pull out the cardinalities. Why not instead just let this function accept a list of integers (the cardinalities themselves)? This encourages unreadable code such as this (repeated throughout the notebooks): `get_embedding_size(proc.stats["categories"], proc.columns_ctx["categorical"]["base"])`

Reply:
@benfred not sure if you're notified without this mention

Reply:
This was built this way in case you changed the column ordering of the preprocessed dataset. We give a dictionary with the encodings and then a list dictating the order of the embeddings. If we agree this level of malleability is not warranted then I am good with only reflecting the embeddings in the order of preprocessing.

Reply:
Oh okay, so the motivation is to avoid mistakes from users who are unaware of the internal reordering. This makes sense 👍

My only comment then is that I think the interface should be clearer / cleaner. Would there be any drawback of adding this functionality within a `Workflow`? To me it makes sense since we are coupling *getting embedding sizes* with *the inherent ordering produced by a workflow*.

I see that right now the `Workflow` has the method `get_final_cols_names`. I could imagine two additional methods, `get_cat_cardinalities` and `get_cat_embedding_sizes`, which wouldn't even need to take any arguments and so would be extremely simple to use?

I'm imagining replacing code like this

```
sizes = get_embedding_size(proc.stats["categories"], proc.columns_ctx["categorical"]["base"])
```

with code like this

```
sizes = proc. get_cat_embedding_sizes()
```

Reply:
At it's core, isn't this function just a mapping from an int cardinality to an int embedding dim? It seems like instead of forcing iteration inside the function, the function should just do that mapping on individual ints and let the user iterate on their calling of it. If we're concerned about the column reordering, it seems more sensible to expose the right ordering via some property and let the user use that to iterate, rather than continuing to bury confusing behavior beneath APIs (eventually someone will have a use case that requires extension, and not knowing about that reordering will break things).

This is probably best for a separate issue, but I do think we'll eventually want to drop the idea of ordering altogether and feed models via dicts like in the KerasSequenceDataset. Besides having the advantage of being cleaner and avoiding these sorts of problems, it will basically be mandatory once we start supporting multi-hot categorical features which can't be cleanly fit into rectangular matrices. This might cost us some performance on the embedding (since we're not doing a single slice on a 2D matrix), but the solution to this should be on the framework side to build an embedding op which can efficiently handle these dict inputs.

Reply:
I thought @jperez999 's point was that this function should safeguard against using a user-defined ordering, because it is inevitable that users will a) not know of the underlying reordering and b) not know that an exposed 'correct' ordering exists, which in turn will lead to strange and hard-to-debug behavior.

That said, I can see either option being reasonable, and either is better than the current approach.

As an avid PyTorch user, I disagree with the *feed models with dicts* sentiment: it feels like a huge step backwards. Why is it mandatory for multi hot? I have not thought about that problem at length but I feel like other options should exist. For example why not use sparse binary tensors?

Reply:
I guess my point is that we can't safeguard against this at every turn, and to a certain extent how and the order in which the user wants to retrieve their embedding sizes is sort of their business (and more exotic models will require more exotic retrieval methods). I would rather make this a known and mandatory part of the API that users need to navigate explicitly, and simplify function calls accordingly, rather than try to anticipate every use case where ordering might be important and head it off with a custom function.

Let's take the dict discussion somewhere else, it probably doesn't make sense here. But I think there are lots of good reasons to prefer it.

Reply:
There are a lot of good points here - but I also think we should keep the conversation going elsewhere. I'm going to merge this PR as is since I think we all agree its a small step in the right direction, even if maybe doesn't go quite far enough. Also the example notebooks are all broken in master, and this change fixes so I think we should get it =).